### PR TITLE
First pass of the CDC Vaccination Indicator

### PR DESCRIPTION
**Pull request by Ananya-Joshi**
### Description
First pass on the CDC Indicator with tests

### CODE REVIEW

**Comment by Ananya-Joshi** [comment]
Missing HI and TX data by county.

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (krivard) [inline cdc_vaccines/delphi_cdc_vaccines/pull.py:95]: this longhand can stay; all the ways I can think of to abstract it out are worse

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (krivard) [inline cdc_vaccines/delphi_cdc_vaccines/pull.py:128]: was this for debugging or should it be logged?

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (krivard) [inline cdc_vaccines/delphi_cdc_vaccines/pull.py:122]: leverage regularity here, eg add to constants:
```
DIFFERENCE_MAPPING = {
    f"{INCIDENCE}_counts_{completeness}_vaccine{age}": f"{CUMULATIVE}_counts_{completeness}_vaccine{age}"
    for completeness, age in product(COMPLETENESS, AGE)
}
SIGNALS = list(DIFFERENCE_MAPPING.keys()) + list(DIFFERENCE_MAPPING.values())
```

then this section becomes
```
for to, from in DIFFERENCE_MAPPING.items():
    df[to] = df[from].diff()
```

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (krivard) [inline cdc_vaccines/delphi_cdc_vaccines/pull.py:161]: would this work?
```suggestion
        # Reorder
        ["fips", "timestamp"] + SIGNALS,
```

File: cdc_vaccines/delphi_cdc_vaccines/run.py
Comment (krivard) [inline cdc_vaccines/delphi_cdc_vaccines/run.py:58]: style nit
```suggestion
            all_data[['timestamp','fips', sensor]],
            from_col='fips',
```

File: cdc_vaccines/delphi_cdc_vaccines/constants.py
Comment (krivard) [inline cdc_vaccines/delphi_cdc_vaccines/constants.py:46]: if this is just being used as a geomapper input, it's fine, but if it is also getting used as an input to `export_csvs` then it has to be `state` or the API will reject the file. I think we typically use `state` here in the constants file and conditionally use `state_code` only when we pass it to geomapper.

File: cdc_vaccines/README.md
Comment (krivard) [inline cdc_vaccines/README.md:5]:
```suggestion
We also aggregate the data to the MSA, HRR, State, HHS Region, and Nation levels. 
```

File: cdc_vaccines/README.md
Comment (krivard) [inline cdc_vaccines/README.md:4]:
```suggestion
daily vaccination rates data as-is, and publish the result as a COVIDcast signal. 
```

File: cdc_vaccines/README.md
Comment (krivard) [inline cdc_vaccines/README.md:3]:
```suggestion
This indicator provides the official vaccination counts in the US. We export the county-level 
```

File: cdc_vaccines/README.md
Comment (krivard) [inline cdc_vaccines/README.md:10]: "There are unknown counties in each state and a row for unknown county and unknown state." 

Can you elaborate on this?

File: cdc_vaccines/params.json.template
Comment (krivard) [inline cdc_vaccines/params.json.template:34]: We typically don't put archiving parameters in the developer params template, and we _definitely_ don't want to overwrite the usafacts archive cache.
```suggestion
```

**Review by krivard — CHANGES_REQUESTED** [review]
Excellent start!

File: cdc_vaccines/README.md
Comment (Ananya-Joshi) [reply cdc_vaccines/README.md:10]: Sorry it wasn't clear! Here is an image from the raw data that better shows what is happening: 
Screen Shot 2021-09-10 at 3 06 59 PM
These are the unknown counties in each state, and the last row here is an unknown county in an unknown state. They have non-zero values for vaccination.

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (Ananya-Joshi) [reply cdc_vaccines/delphi_cdc_vaccines/pull.py:128]: Removed!

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (dshemetov) [reply cdc_vaccines/delphi_cdc_vaccines/pull.py:31]: Sure I'll take a look on Monday!

File: cdc_vaccines/delphi_cdc_vaccines/constants.py
Comment (Ananya-Joshi) [reply cdc_vaccines/delphi_cdc_vaccines/constants.py:42]: Changed! Thank you for the suggestion!

**Comment by Ananya-Joshi** [comment]
The changes requested have been made. We should talk about the two open issues still - the unknown states/counties and the megacounty thread.

File: cdc_vaccines/README.md
Comment (krivard) [reply cdc_vaccines/README.md:10]:
```suggestion
there is no county level data for counties in Texas and Hawaii. Each state has some vaccination counts assigned to "unknown county". Some vaccination counts are assigned to "unknown state, unknown county".
```

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (dshemetov) [reply cdc_vaccines/delphi_cdc_vaccines/pull.py:30]: Alright, looking through our chats on prop signals, this is what can happen: if megacounties are used to keep track of unassigned counts, then we must be careful not to add the megacounty population counts when doing prop signals, otherwise when aggregating to larger geos we will double count the state population. I don't see any prop signals atm, so this seems fine.

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (dshemetov) [inline cdc_vaccines/delphi_cdc_vaccines/pull.py:97]: What does the 00000 FIPS code represent? Not assigned to any county or state?

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (dshemetov) [inline cdc_vaccines/delphi_cdc_vaccines/pull.py:115]: Btw, you might like [this](https://github.com/cmu-delphi/covidcast-indicators/blob/b4c9b5573232b89c45cf3ce8ea6d0bf61ef53133/jhu/delphi_jhu/pull.py#L27) version of taking diffs, grouped by geos.

Comment (dshemetov) [reply cdc_vaccines/delphi_cdc_vaccines/pull.py:115]: I think the `.isna()` in that code should probably be replaced with `.loc[min_time_value, :]` or something, but the gist is there.

File: cdc_vaccines/delphi_cdc_vaccines/run.py
Comment (dshemetov) [inline cdc_vaccines/delphi_cdc_vaccines/run.py:75]: Also, if you anticipate any nans in the dataset that will be ingested into the db, consider annotating them with nancodes in the missing columns, like [here](https://github.com/dshemetov/covidcast-indicators/blob/263485b8d4d94a36ebd9ed6292d29b62b69efc64/jhu/delphi_jhu/run.py#L67).

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (Ananya-Joshi) [reply cdc_vaccines/delphi_cdc_vaccines/pull.py:97]: I think it refers to when the state and the county are both listed as "UNK". So yes, it only gets used at aggregate at the nation level. What do you think?

File: cdc_vaccines/delphi_cdc_vaccines/run.py
Comment (Ananya-Joshi) [reply cdc_vaccines/delphi_cdc_vaccines/run.py:75]: Should I remove these columns? I kept them because they were in the USAFacts indicator, but I can remove them if needed.

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (Ananya-Joshi) [reply cdc_vaccines/delphi_cdc_vaccines/pull.py:115]: I wonder if we can keep this method for now, but then later I'll look to fix the method and use it.

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (dshemetov) [reply cdc_vaccines/delphi_cdc_vaccines/pull.py:97]: That makes sense, yea. Just curious about the data! Not much to do here.

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (dshemetov) [reply cdc_vaccines/delphi_cdc_vaccines/pull.py:115]: Sure, that's alright. A possible refactor for later.

File: cdc_vaccines/delphi_cdc_vaccines/run.py
Comment (dshemetov) [reply cdc_vaccines/delphi_cdc_vaccines/run.py:75]: Oh no, I mean that we now support three extra columns `missing_value`, `missing_stderr`, and `missing_sample_size`, where we write nancodes for why a value is missing. See the link for an example. The possible nancodes are [here](https://github.com/dshemetov/covidcast-indicators/blob/263485b8d4d94a36ebd9ed6292d29b62b69efc64/_delphi_utils_python/delphi_utils/nancodes.py). So for example, this indicator is not a sample-based indicator so stderr and sample_size just do not apply, so we would write `Nans.NOT_APPLICABLE` in the corresponding columns.

Comment (krivard) [reply cdc_vaccines/delphi_cdc_vaccines/run.py:75]: Since this is a new indicator, we should definitely be creating it with the correct Nans codes in the missing_* columns as needed. It sounds like this indicator should use:
* `missing_stderr = Nans.NOT_APPLICABLE`
* `missing_sample_size = Nans.NOT_APPLICABLE`

File: cdc_vaccines/delphi_cdc_vaccines/pull.py
Comment (Ananya-Joshi) [reply cdc_vaccines/delphi_cdc_vaccines/pull.py:30]: Okay! So from what I understand, I'll leave as in for now!

File: cdc_vaccines/delphi_cdc_vaccines/run.py
Comment (dshemetov) [inline cdc_vaccines/delphi_cdc_vaccines/run.py:72]: This is more what I meant
```suggestion
        df["se"] = np.nan
        df["sample_size"] = np.nan
        df["missing_se"] = nancodes.Nans.NOT_APPLICABLE
        df["missing_sample_size"] = nancodes.Nans.NOT_APPLICABLE
```

Comment (Ananya-Joshi) [reply cdc_vaccines/delphi_cdc_vaccines/run.py]: Sorry! I messed up version control. Thanks for the catch - it should all be taken care of now :)

Comment (Ananya-Joshi) [reply cdc_vaccines/delphi_cdc_vaccines/run.py:72]: It looks like create_export_csv in delphi_utils needs the 'se' and 'sample_size' columns - not the missing columns - should I change the delphi utils function? 

../../_delphi_utils_python/delphi_utils/export.py:73: in create_export_csv
    export_df = df[df["timestamp"] == date][["geo_id", "val", "se", "sample_size",]]

Comment (Ananya-Joshi) [reply cdc_vaccines/delphi_cdc_vaccines/run.py]: Okay, I'll wait until that gets merged in.

**Comment by Ananya-Joshi** [comment]
@krivard @dshemetov , do we need to do anything else for this indicator?

File: cdc_vaccines/delphi_cdc_vaccines/run.py
Comment (dshemetov) [reply cdc_vaccines/delphi_cdc_vaccines/run.py]: Now that #1252 has been merged, my suggestion should work

**Comment by dshemetov** [comment]
@Ananya-Joshi just the Nancodes annotation in the thread above and it should be good to go 👍

**Comment by Ananya-Joshi** [comment]
> @Ananya-Joshi just the Nancodes annotation in the thread above and it should be good to go 👍
As far as I can see on my local, they're changed! Is this not the case for what you can see?

**Comment by dshemetov** [comment]
Oh, now we're missing the actual "se" and "sample_size" columns. I will make a commit with what I mean.

**Review by dshemetov — APPROVED** [review]
LGTM

**Comment by krivard** [comment]
This looks like rebase leavings:
image

Indeed, the gitref for "Remove test for old export func" is supposed to be 5fa63149d and not ca09586. @Ananya-Joshi can you clarify what was done here? I want to make sure when we merge it we won't wind up with duplicate copies of all of those dual-author commits.

**Pull request closed without merging**
